rllib/core/rl_trainer/torch/torch_rl_trainer.py

```diff
@@ -1,10 +1,185 @@
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+        loss[self.TOTAL_LOSS_KEY].backward()
+        grads = {pid: p.grad for pid, p in self._params.items()}
+
+        return grads
+
+    @override(RLTrainer)
+    def apply_gradients(self, gradients: ParamDictType) -> None:
+
+        # make sure the parameters do not carry gradients on their own
+        for optim in self._optim_to_param:
+            optim.zero_grad(set_to_none=True)
+
+        # set the gradient of the parameters
+        for pid, grad in gradients.items():
+            self._params[pid].grad = grad
+
+        # for each optimizer call its step function with the gradients
+        for optim in self._optim_to_param:
+            optim.step()
+
+    @override(RLTrainer)
+    def build(self) -> None:
+        # TODO (Kourosh): How do we handle model parallism?
+        # TODO (Kourosh): Instead of using _TorchAccelerator, we should use the public
+        # api in ray.train but allow for session to be None without any errors raised.
+        if self._use_gpu:
+            self._device = _TorchAccelerator().get_device()
+        else:
+            self._device = torch.device("cpu")
+        super().build()
+
+    @override(RLTrainer)
+    def _make_distributed(self) -> MultiAgentRLModule:
+        module = self._make_module()
+
+        # if the module is a MultiAgentRLModule and nn.Module we can simply assume
+        # all the submodules are registered. Otherwise, we need to loop through
+        # each submodule and move it to the correct device.
+        # TODO (Kourosh): This can result in missing modules if the user does not
+        # register them in the MultiAgentRLModule. We should find a better way to
+        # handle this.
+        if isinstance(module, torch.nn.Module):
+            module.to(self._device)
+            module = DDPRLModuleWrapper(module)
+        else:
+            for key in module.keys():
+                module[key].to(self._device)
+                module[key] = DDPRLModuleWrapper(module[key])
+
+        return module
+
+    @override(RLTrainer)
+    def _convert_batch_type(self, batch: MultiAgentBatch):
+        batch = NestedDict(batch.policy_batches)
+        batch = NestedDict(
+            {
+                k: torch.as_tensor(v, dtype=torch.float32, device=self._device)
+                for k, v in batch.items()
+            }
+        )
+        return batch
+
+    @override(RLTrainer)
+    def do_distributed_update(self, batch: MultiAgentBatch) -> Mapping[str, Any]:
+        # in torch the distributed update is no different than the normal update
+        return self._update(batch)
+
+    @override(RLTrainer)
+    def get_param_ref(self, param: ParamType) -> Hashable:
+        return param
+
+    @override(RLTrainer)
+    def get_parameters(self, module: RLModule) -> Sequence[ParamType]:
+        return list(module.parameters())
+
+    @override(RLTrainer)
+    def get_optimizer_obj(
```

Comment:
Rename to `make_optimizer` to remain analogous to `_make_module` and `_make_distributed_module` (see my comment for that method as well).
Logic:
`make_...` -> creates a new object for this trainer
`get_...` -> some getter (or light weight data compilation step, e.g. config sub-objects or `RLTrainer.get_parameters()`)